### PR TITLE
[app_dart] Add swarming authentication provider

### PR DESCRIPTION
**Pull request by CaseyHillers**
Add a `SwarmingAuthenticationProvider` that is a child of the existing `AuthenticationProvider`. This provider only authenticates for Cocoon agents and LUCI swarming bot service accounts.

This ensures a given `RequestHandler` can only be updated through these methods (enforcing an ACL). For now, I added this to the `update-task-status`, but we should do a greater security review to audit endpoints that should only be used by automated systems.

Added LUCI swarming authentication which reuses the existing Google authentication.

# Issues
https://github.com/flutter/flutter/issues/66191

# Tests

- Added tests that ensure this child class does not grant the same permissions as the parent
- Added tests for LUCI auth being passed to id token

### CODE REVIEW

**Comment by keyonghan** [comment]
why not expanding existing `AuthenticationProvider`, but creating a separate `SwarmingAuthenticationProvider`?

**Comment by CaseyHillers** [comment]
> why not expanding existing `AuthenticationProvider`, but creating a separate `SwarmingAuthenticationProvider`?

1. This improves the security of `update-task-status` so only automated systems can push changes. Right now, any dashboard user can push changes.
2. `AuthenticationProvider` has 4 separate logic branches that are hard to follow to ensure authentication. Adding a 5th is somewhat trivial, but it adds more complexity. It gets hard to ensure good test coverage of this provider when it has so many different paths.

File: app_dart/lib/src/request_handling/authentication.dart [hunk 1]
```diff
@@ -60,6 +60,7 @@ import 'exceptions.dart';
 /// See also:
 ///
 ///  * <https://cloud.google.com/appengine/docs/standard/python/reference/request-response-headers>
+// TODO(chillers): Remove (1) when DeviceLab has migrated to LUCI, https://github.com/flutter/flutter/projects/151#card-47536851
```
Comment (keyonghan) [inline app_dart/lib/src/request_handling/authentication.dart:63]: With the same logic being implemented in `SwarmingAuthenticationProvider`, can we delete this part now?

File: app_dart/lib/src/request_handling/swarming_authentication.dart [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+      return await authenticateIdToken(swarmingToken, clientContext: clientContext, log: log);
+    }
+
+    throw const Unauthenticated('Request rejected due to not from LUCI or Cocoon agent');
+  }
+
+  // This method is expensive (run time of ~1,500ms!). If the server starts
+  // handling any meaningful API traffic, we should move request processing
+  // to dedicated isolates in a pool.
+  static bool _compareHashAndPassword(List<int> serverAuthTokenHash, String clientAuthToken) {
```
Comment (keyonghan) [inline app_dart/lib/src/request_handling/swarming_authentication.dart:125]: move to a common script so both `SwarmingAuthenticationProvider` and `Authentication` can share?

Comment (CaseyHillers) [reply app_dart/lib/src/request_handling/swarming_authentication.dart]: Done. Made public in `AuthenticationProvider`, with plan to remove after devicelab luci migration is complete.

File: app_dart/lib/src/request_handling/authentication.dart (hunk 1 above)
Comment (CaseyHillers) [reply app_dart/lib/src/request_handling/authentication.dart:63]: `/api/append-log`, `/api/get-authentication-status`, and `'/api/update-agent-health'` still use the `AuthenticationProvider`, so not yet. I'll investigate what endpoints require what auth methods, and update in a separate PR.

My expectation is for this to be done after the devicelab to LUCI migration as we will be able to remove several request handlers.

Comment (keyonghan) [reply app_dart/lib/src/request_handling/authentication.dart:63]: SGTM

**Review by keyonghan — APPROVED** [review]
LGTM
